File: apps/admin/src/api/procurement.ts
import { secured } from './client';

export type Supplier = {
  id: number;
  name: string;
  contactName: string;
  phone: string;
  email: string;
  address: string;
  status: 'active' | 'inactive';
};

export type SupplierInput = Omit<Supplier, 'id'>;

export type Branch = { id: number; name: string; code: string };

export type PurchaseOrderItem = {
  id: number;
  inventoryItemId: number;
  name: string;
  quantityOrdered: number;
  quantityReceived: number;
  unit: string;
  unitCost: number;
};

export type PurchaseOrder = {
  id: number;
  branchId: number;
  branchName: string;
  supplierId: number;
  supplierName: string;
  status:
    | 'draft'
    | 'submitted'
    | 'approved'
    | 'ordered'
    | 'partially_received'
    | 'received'
    | 'cancelled';
  note: string;
  createdAt: string;
  items: PurchaseOrderItem[];
};

export type StockMovement = {
  id: number;
  inventoryItemId: number;
  inventoryItemName: string;
  movementType:
    | 'initial'
    | 'purchase_receipt'
    | 'stock_request_receipt'
    | 'pos_sale'
    | 'adjustment';
  quantityDelta: number;
  quantityBefore: number;
  quantityAfter: number;
  note: string;
  createdAt: string;
};

export const listSuppliers = () => secured<Supplier[]>('/suppliers');
export const createSupplier = (data: SupplierInput) =>
  secured<{ id: number }>('/suppliers', { method: 'POST', data });
export const updateSupplier = (id: number, data: SupplierInput) =>
  secured<{ id: number }>(`/suppliers/${id}`, { method: 'PATCH', data });
export const listBranches = () => secured<Branch[]>('/branches');
export const listPurchaseOrders = () =>
  secured<PurchaseOrder[]>('/purchase-orders');
export const createPurchaseOrder = (data: {
  branchId: number;
  supplierId: number;
  note: string;
  items: { inventoryItemId: number; quantity: number; unitCost: number }[];
}) =>
  secured<{ id: number; status: 'draft' }>('/purchase-orders', {
    method: 'POST',
    data,
  });
export const updatePurchaseOrderStatus = (
  id: number,
  status: 'submitted' | 'approved' | 'ordered' | 'cancelled',
) =>
  secured<{ id: number; status: PurchaseOrder['status'] }>(
    `/purchase-orders/${id}/status`,
    { method: 'PATCH', data: { status } },
  );
export const receivePurchaseOrder = (
  id: number,
  items: { itemId: number; quantity: number }[],
  note = '',
) =>
  secured<{ id: number; status: PurchaseOrder['status'] }>(
    `/purchase-orders/${id}/receive`,
    { method: 'POST', data: { items, note } },
  );
export const listStockMovements = (branchCode: string) =>
  secured<StockMovement[]>(
    `/stock-movements?branchCode=${encodeURIComponent(branchCode)}`,
  );
export const adjustInventory = (
  id: number,
  quantity: number,
  note: string,
  branchCode: string,
) =>
  secured<{ id: number; quantity: number }>(
    `/inventory/${id}/adjust?branchCode=${encodeURIComponent(branchCode)}`,
    { method: 'POST', data: { quantity, note } },
  );

File: apps/api/internal/dto/procurement.go
package dto

type SupplierRequest struct {
	Name        string `json:"name" binding:"required"`
	ContactName string `json:"contactName"`
	Phone       string `json:"phone"`
	Email       string `json:"email"`
	Address     string `json:"address"`
	Status      string `json:"status" binding:"omitempty,oneof=active inactive"`
}

type PurchaseOrderItemRequest struct {
	InventoryItemID int64   `json:"inventoryItemId" binding:"required,gt=0"`
	Quantity        float64 `json:"quantity" binding:"required,gt=0"`
	UnitCost        float64 `json:"unitCost" binding:"gte=0"`
}

type PurchaseOrderRequest struct {
	BranchID   *int64                     `json:"branchId"`
	SupplierID int64                      `json:"supplierId" binding:"required,gt=0"`
	Note       string                     `json:"note"`
	Items      []PurchaseOrderItemRequest `json:"items" binding:"required,min=1,dive"`
}

type PurchaseOrderStatusRequest struct {
	Status string `json:"status" binding:"required,oneof=submitted approved ordered cancelled"`
}

type PurchaseOrderReceiptItemRequest struct {
	ItemID   int64   `json:"itemId" binding:"required,gt=0"`
	Quantity float64 `json:"quantity" binding:"required,gt=0"`
}

type PurchaseOrderReceiptRequest struct {
	Note  string                            `json:"note"`
	Items []PurchaseOrderReceiptItemRequest `json:"items" binding:"required,min=1,dive"`
}

type StockAdjustmentRequest struct {
	Quantity float64 `json:"quantity" binding:"required,gte=0"`
	Note     string  `json:"note" binding:"required,max=500"`
}
